A80-tcp-telemetry.md

```diff
@@ -0,0 +1,114 @@
+| grpc.tcp.sender\_latency | Histogram (floating-point) | {s} | None | Time taken by the TCP socket to write the first byte of a write onto the NIC. This includes the latency incurred by traffic shaping, qdisc, throttling, and pacing at the sender. Corresponds to the time taken between the final `SCHED` timestamp and the `SENT` timestamp. Sampled periodically. |
+| grpc.tcp.transfer\_latency | Histogram (floating-point) | {s} | size (Bytes)<sup>1</sup> | Time taken to transmit the first size bytes of a write. Transfer latency is measured from when the first byte is handed to the NIC until TCP receives the acknowledgement for the last byte. Corresponds to the time taken between the `SENT` timestamp and the `ACKED` timestamp. Sampled periodically. |
+
+<sup>1</sup> Since transfer latency is strongly affected by the write size, it is broken down into different buckets based on the size of the write. Further, we will only measure the latencies of certain benchmark sample sizes to get measurements that are unaffected by the write sizes. The proposed buckets are:
+
+| Buffered Size | Benchmark Size | `Size (Bytes)` |
+| :---- | :---- | :---- |
+| \[0, 1KiB) | Whole buffer | 1024 |
+| \[1KiB, 8KiB) | First 1KiB | 1024 |
+| \[8KiB, 64KiB) | First 8KiB | 8196 |
+| \[64KiB, 256KiB) | First 64KiB | 65536 |
+| \[256KiB, 2MiB) | First 256KiB | 262144 |
+| \[2MiB, \+inf) | First 2MiB | 2097152 |
+
+#### Suggested Metric Collection Algorithm
+
+* Set TCP\_LATENCY\_RECORD\_FREQUENCY to a default of 1 in 1000 writes.  
+  * Implementations can choose to make it configurable.  
+* For each new connected TCP socket,   
+  * Set `writes_since_last_latency_measurement_` to a random integer in \[0, TCP\_LATENCY\_RECORD\_FREQUENCY). Increment this value for every write.  
+  * Perform any prerequisites needed for the socket to support timestamping.   
+    * For Linux TCP, this involves making a setsockopt(SO\_TIMESTAMPING) call with the flag value set to SOF\_TIMESTAMPING\_SOFTWARE | SOF\_TIMESTAMPING\_OPT\_ID | SOF\_TIMESTAMPING\_OPT\_TSONLY | SOF\_TIMESTAMPING\_OPT\_ID\_TCP | SOF\_TIMESTAMPING\_OPT\_STATS.  
+* If `writes_since_last_latency_measurement_` % TCP\_LATENCY\_RECORD\_FREQUENCY \= 0  
+  * Enable latency measurement for the write.   
+    * For Linux TCP, this involves splitting the write into two chunks based on the buckets listed above and adding a SO\_TIMESTAMPING cmsg header to the sendmsg call with the flags SOF\_TIMESTAMPING\_TX\_SCHED | SOF\_TIMESTAMPING\_TX\_SOFTWARE on the sampled chunk and SOF\_TIMESTAMPING\_TX\_ACK on the other chunk.   
+    * There is only one chunk if the write is less than 1024 bytes, in that case all three flags are set on this chunk.  
+  * Set `writes_since_last_latency_measurement_` to 1 and repeat.
+
+### Metric Stability
+All metrics added in this proposal will start as experimental. The long term goal will be to de-experimentalize them and have them be on by default, but the exact criteria for that change are TBD.
```

Comment:
Is it our goal to have these on by default? The per-connection metrics are sorta hard to use without the labels, but those have too high a cardinality to be generally useful.